pkg/controller/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -0,0 +1,225 @@
+	t.Helper()
+
+	name := toNamespaceName(username, tcNamespace.Type)
+	namespace := &corev1.Namespace{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: name}, namespace)
+	assert.NoError(t, err)
+	assert.Equal(t, username, namespace.Labels["owner"])
+	assert.Equal(t, tcNamespace.Revision, namespace.Labels["revision"])
+}
+
+func createNamespaceWithLabels(t *testing.T, client client.Client, tcNamespace toolchainv1alpha1.Namespace, username string) {
+	params := make(map[string]string)
+	params["USER_NAME"] = username
+	err := applyTemplateTestMock(client, tcNamespace, params)
+	require.NoError(t, err)
+
+	name := toNamespaceName(username, tcNamespace.Type)
+	namespace := &corev1.Namespace{}
+	err = client.Get(context.TODO(), types.NamespacedName{Name: name}, namespace)
+	require.NoError(t, err)
+
+	namespace.Labels = make(map[string]string)
+	namespace.Labels["owner"] = username
+	namespace.Labels["revision"] = tcNamespace.Revision
+
+	err = client.Update(context.TODO(), namespace)
+	require.NoError(t, err)
+}
+
+func newNSTmplSet(userName string) *toolchainv1alpha1.NSTemplateSet {
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      userName,
+			Namespace: namespaceName,
+		},
+		Spec: toolchainv1alpha1.NSTemplateSetSpec{
+			TierName: "basic",
+			Namespaces: []toolchainv1alpha1.Namespace{
+				{Type: "dev", Revision: "rev1", Template: ""},
+				{Type: "code", Revision: "rev1", Template: ""},
+			},
+		},
+	}
+	return nsTmplSet
+}
+
+func prepareReconcile(t *testing.T, username string, initObjs ...runtime.Object) (*ReconcileNSTemplateSet, reconcile.Request, *test.FakeClient) {
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+	fakeClient := test.NewFakeClient(t, initObjs...)
+
+	r := &ReconcileNSTemplateSet{
+		client:        fakeClient,
+		scheme:        s,
+		applyTemplate: applyTemplateTestMock,
+	}
+	return r, newReconcileRequest(username), fakeClient
+}
+
+func newReconcileRequest(name string) reconcile.Request {
+	return reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespaceName,
+		},
+	}
+}
+
+// applyTemplate test mock
+func applyTemplateTestMock(client client.Client, tcNamespace toolchainv1alpha1.Namespace, params map[string]string) error {
+	userName := params["USER_NAME"]
+	name := toNamespaceName(userName, tcNamespace.Type)
+
+	// create project
+	project := &projectv1.Project{}
+	if err := client.Get(context.TODO(), types.NamespacedName{Name: name}, project); err != nil {
+		if errors.IsNotFound(err) {
+			prjReq := &projectv1.Project{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: name,
+				},
+				Status: projectv1.ProjectStatus{
+					Phase: corev1.NamespaceActive,
+				},
+			}
+			if err := client.Create(context.TODO(), prjReq); err != nil {
+				return err
+			}
+		} else {
```

Comment:
can we remove the else block and just return if the `if` condition isn't met?

Reply:
If creation of project is successful (line-196) then we should go ahead with creation of namespace (line-205 onward) and for the same we need else block here.

Reply:
@nurali-techie this else pertains to the if errors.IsNotFound(err) {...} block right? 

```
if errors.IsNotFound(err) {
			prjReq := &projectv1.Project{
				ObjectMeta: metav1.ObjectMeta{
					Name: name,
				},
				Status: projectv1.ProjectStatus{
					Phase: corev1.NamespaceActive,
				},
			}
			if err := client.Create(context.TODO(), prjReq); err != nil {
				return err
			}
		} else {
			return err
		}
```
we shouldn't need the else here.. unless i'm missing something?